python/ray/tune/utils/object_cache.py

```diff
@@ -0,0 +1,165 @@
+        """Decrease number of max objects for this key.
+
+        Args:
+            key: Group key.
+            by: Decrease by this amount.
+        """
+        self._max_num_objects[key] -= by
+
+    def has_cached_object(self, key: T) -> bool:
+        """Return True if at least one cached object exists for this key.
+
+        Args:
+            key: Group key.
+
+        Returns:
+            True if at least one cached object exists for this key.
+        """
+        return bool(self._cached_objects[key])
+
+    def cache_object(self, key: T, obj: U) -> bool:
+        """Cache object for a given key.
+
+        This will put the object into a cache, assuming the number
+        of cached objects for this key is less than the number of
+        max objects for this key.
+
+        An exception is made if eager caching is enaabled and no other
+        objects are cached globally. In that case, the object can
+        still be cached.
+
+        Args:
+            key: Group key.
+            obj: Object to cache.
+
+        Returns:
+            True if the object has been cached. False otherwise.
+        """
+        # If we have more objects cached already than we desire
+        if len(self._cached_objects[key]) >= self._max_num_objects[key]:
+            # If eager caching is disabled, never cache
+            if not self._eager_caching:
+                return False
+
+            # If we have more than one other cached object, don't cache
+            if self._num_cached_objects > 0:
+                return False
+
+            # If any other objects are expected to be cached, don't cache
+            if any(v for v in self._max_num_objects.values()):
+                return False
+
+        # Otherwise, cache (for now).
+
+        self._cached_objects[key].append(obj)
+        self._num_cached_objects += 1
+        return True
+
+    def pop_cached_object(self, key: T) -> Optional[U]:
+        """Get one cached object for a key.
+
+        This will remove the object from the cache.
+
+        Args:
+            key: Group key.
+
+        Returns:
+            Cached object.
+        """
+        if not self.has_cached_object(key):
+            return None
+
+        self._num_cached_objects -= 1
+        return self._cached_objects[key].pop(0)
+
+    def flush_cached_objects(self, force_all: bool = False) -> Generator[U, None, None]:
+        """Return a generator over cached objects evicted from the cache.
+
+        This method yields all cached objects that should be evicted from the
+        cache for cleanup by the caller.
+
+        If the number of max objects is lower than the number of
+        cached objects for a given key, objects are evicted until
+        the numbers are equal.
+
+        If ``keep_one=True``, one (global) object is allowed to remain cached.
+
+        If ``force_all=True``, all objects are evicted.
+
+        Args:
+            keep_one: If True, may keep one object cached, even if the number
```

Comment:
We should also mention that this will keep the latest one, of the last key.

```
A(max=0) | a b
B(max=0) | c
C(max=0) | d e f
```

```
A(max=0) | 
B(max=0) | 
C(max=0) | f
```